plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java

```diff
@@ -32,16 +34,61 @@
 import org.eclipse.jface.wizard.IWizard;
 import org.eclipse.m2e.core.MavenPlugin;
 import org.eclipse.m2e.core.internal.IMavenConstants;
+import org.eclipse.m2e.core.internal.lifecyclemapping.discovery.LifecycleMappingDiscoveryRequest;
 import org.eclipse.m2e.core.internal.project.ProjectConfigurationManager;
 import org.eclipse.m2e.core.project.IProjectConfigurationManager;
 import org.eclipse.m2e.core.project.LocalProjectScanner;
 import org.eclipse.m2e.core.project.MavenProjectInfo;
 import org.eclipse.m2e.core.project.MavenUpdateRequest;
 import org.eclipse.m2e.core.project.ResolverConfiguration;
+import org.eclipse.m2e.core.ui.internal.wizards.LifecycleMappingDiscoveryHelper;
+import org.eclipse.m2e.core.ui.internal.wizards.MappingDiscoveryJob;
+import org.eclipse.ui.internal.wizards.datatransfer.EasymportJob;
 import org.eclipse.ui.wizards.datatransfer.ProjectConfigurator;
 
 public class MavenProjectConfigurator implements ProjectConfigurator {
 	
+	private static class CumulativeMappingDiscoveryJob extends MappingDiscoveryJob {
+		private static CumulativeMappingDiscoveryJob INSTANCE;
```

Comment:
The constraints are:
* MavenProjectConfigurator is stateless (its lifecycle and the multiplicity aren't specified, and I'd rather avoid to constrain them). 
* UpdateMavenConfigurationJob has multiple instances whom orchestration isn't something we can control
* While importing, we don't want the Mapping discovery dialog to pop-up before the import isn't completed, so we have to make sure only one instance will be used for an import operation. 

Basically, we need a single stateful operation, after multiple stateless ones. The singleton job and a join on the family is the best thing I found to allow that.